app/src/test/java/com/github/se/signify/model/user/UserViewModelTest.kt

```diff
@@ -120,4 +141,380 @@ class UserViewModelTest {
+    val exception = Exception("Failed to update profile picture")
+    doAnswer {
+          val onFailure = it.getArgument<(Exception) -> Unit>(3)
+          onFailure(exception)
+          null
+        }
+        .whenever(userRepository)
+        .updateProfilePictureUrl(eq(currentUserId), eq(uri), any(), any())
+
+    // Act
+    userViewModel.updateProfilePictureUrl(uri)
+
+    // Assert
+    verify(userRepository).updateProfilePictureUrl(eq(currentUserId), eq(uri), any(), any())
+  }
+
+  @Test
+  fun getOngoingChallengesUpdatesChallengesOnSuccess() = runTest {
+    // Arrange
+    val challenges =
+        listOf(Challenge(challengeId = "challenge1"), Challenge(challengeId = "challenge2"))
+    doAnswer {
+          val onSuccess = it.getArgument<(List<Challenge>) -> Unit>(1)
+          onSuccess(challenges)
+          null
+        }
+        .whenever(userRepository)
+        .getOngoingChallenges(eq(currentUserId), any(), any())
+
+    // Act
+    userViewModel.getOngoingChallenges()
+
+    // Assert
+    assertEquals(challenges, userViewModel.ongoingChallenges.value)
+  }
+
+  @Test
+  fun getOngoingChallengesHandlesFailure() = runTest {
+    // Arrange
+    val exception = Exception("Failed to fetch ongoing challenges")
+    doAnswer {
+          val onFailure = it.getArgument<(Exception) -> Unit>(2)
+          onFailure(exception)
+          null
+        }
+        .whenever(userRepository)
+        .getOngoingChallenges(eq(currentUserId), any(), any())
+
+    // Act
+    userViewModel.getOngoingChallenges()
+
+    // Assert
+    verify(userRepository).getOngoingChallenges(eq(currentUserId), any(), any())
+  }
+
+  @Test
+  fun removeOngoingChallengeHandlesFailure() = runTest {
+    // Arrange
+    val exception = Exception("Failed to remove challenge")
+    doAnswer {
+          val onFailure = it.getArgument<(Exception) -> Unit>(3)
+          onFailure(exception)
+          null
+        }
+        .whenever(userRepository)
+        .removeOngoingChallenge(eq(currentUserId), eq(challengeId), any(), any())
+
+    // Act
+    userViewModel.removeOngoingChallenge(currentUserId, challengeId)
+
+    // Assert
+    verify(userRepository).removeOngoingChallenge(eq(currentUserId), eq(challengeId), any(), any())
+  }
+
+  @Test
+  fun checkAndUnlockNextQuestSetsInitialAccessDateWhenNull() = runTest {
+    // Arrange
+    doAnswer {
+          val onSuccess = it.getArgument<(String?) -> Unit>(1)
+          onSuccess(null)
+          null
+        }
+        .whenever(userRepository)
+        .getInitialQuestAccessDate(eq(currentUserId), any(), any())
+
+    doAnswer {
+          val onSuccess = it.getArgument<() -> Unit>(2)
+          onSuccess()
+          null
+        }
+        .whenever(userRepository)
+        .setInitialQuestAccessDate(eq(currentUserId), any(), any(), any())
+
+    // Act
+    userViewModel.checkAndUnlockNextQuest()
+
+    // Assert
+    assertEquals("1", userViewModel.unlockedQuests.value)
```

Comment:
Adding comments explaining the purpose of each assertion and the expected state after each action can make the tests easier to follow for other team members.